### DISCUSSION TITLE
SSG cache headers - what to set?

### DISCUSSION THREAD

**Discussion by MrMarci666** [vercel/next.js#70010] [Help]
### Summary

We were using Gatsby for years for rendering websites, the flow is the following:

1. In the CMS, the user makes a change
2. This triggers a Job in the CI / CD software, which starts a node container and runs `yarn build` command to generate a static HTML website
3. We upload the generated static HTML, js, css, json, etc... files to a static hosting, like S3 on AWS or an nginx webserver.

This way we were able to create a very secure, fast website without any backend, which is pretty nice in terms of security.
However, now we changed Gatsby to Next.JS, but unfortunately we are not able to find **which cache headers should we apply on which files after generated a site using Next.JS**. We are ONLY using SSG, not using any dynamic content, and NOT running the Next.JS webserver to serve the site!

Gatsby has a clear guideline on this: https://www.gatsbyjs.com/docs/how-to/previews-deploys-hosting/caching/

We would need something familiar. Could you help us? Thank you!

Additionally, I added the current nginx cache configs for our sites. 

### Additional information

```js
location ~* \.(js|css|png|jpg|jpeg|gif|ico)$ {
    expires max;
    log_not_found off;
  }

  # Cache headers based on GatsbyJS documentation
  location ~* \.(?:ico|gif|jpe?g|png|mp4|webp|svg|woff2?|ttf|eot)$ {
    expires 1y;
    add_header Cache-Control "public, max-age=31536000, immutable";
  }

  # Cache headers for HTML files
  location ~* \.html$ {
    add_header Cache-Control "public, max-age=0, must-revalidate";
  }

  # Cache headers for JSON files in the page-data directory
  location ~* /page-data/.*\.json$ {
    add_header Cache-Control "public, max-age=0, must-revalidate";
  }

  # Cache headers for app-data.json
  location = /page-data/app-data.json {
    add_header Cache-Control "public, max-age=0, must-revalidate";
  }

  # Cache headers for static files
  location ^~ /static/ {
    add_header Cache-Control "public, max-age=31536000, immutable";
  }

  # Cache headers for JavaScript and CSS files
  location ~* \.(js|css)$ {
    add_header Cache-Control "public, max-age=31536000, immutable";
  }

  # Cache headers for /sw.js with exception
  location = /sw.js {
    add_header Cache-Control "public, max-age=0, must-revalidate";
  }
```

**Comment by Juneezee** [comment]
Next.js docs has a section for `Cache-Control`, see https://nextjs.org/docs/app/api-reference/next-config-js/headers#cache-control

> Next.js sets the `Cache-Control` header of `public, max-age=31536000, immutable` for truly immutable assets. It cannot be overridden. These immutable files contain a SHA-hash in the file name, so they can be safely cached indefinitely.

**Reply by MrMarci666** [reply]
Thank you for sharing the Next.js documentation link, I had also come across this article. However, I believe there is a notable gap between what the Next.js documentation offers and what I was expecting. While it mentions automatic caching for immutable assets, it doesn't provide the same level of detail or clarity on how to handle caching headers manually in environments where we don't use the Next.js server, such as with static hosting setups (e.g., S3 or Nginx).

The **Gatsby documentation**, by contrast, is very explicit in guiding developers on which specific cache headers to apply to different file types when deploying to static hosting. For example, Gatsby clearly outlines how to handle cache headers for HTML files, JSON files (especially in page-data directories), JavaScript, and CSS files. This makes it very clear for users like me who are generating static sites and deploying them to environments where manual cache control is required.

In the case of **Next.js**, the documentation mainly focuses on server-rendered or dynamic setups, where Next.js automatically manages these headers. However, when using **static site generation (SSG)** and deploying to a platform like Nginx, we have to manually set cache headers. For instance:
- What cache headers should I apply to the **HTML files** generated by SSG?
- Should I treat **JSON files** the same way as JavaScript and CSS?
- While the hashed static assets have immutable caching applied automatically, I need clear guidance on the rest of the files.

In short, while **Next.js handles immutable assets well**, it lacks the detailed, per-file-type guidance that Gatsby provides, which is crucial for developers deploying to static environments. A similar approach to what Gatsby offers would be extremely helpful in such scenarios.
